logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java

```diff
@@ -314,18 +321,32 @@ public long write(Queueable element) throws IOException {
 
             // create a new head page if the current does not have sufficient space left for data to be written
             if (! this.headPage.hasSpace(data.length)) {
-                // beheading includes checkpoint+fsync if required
-                TailPage tailPage = this.headPage.behead();
 
-                this.tailPages.add(tailPage);
-                if (! tailPage.isFullyRead()) {
-                    this.unreadTailPages.add(tailPage);
+                // TODO: verify queue state integrity WRT Queue.open()/recover() at each step of this process
+
+                int newHeadPageNum = this.headPage.pageNum + 1;
+
+                if (this.headPage.isFullyAcked()) {
+                    // purge the old headPage because its full and fully acked
+                    // there is no checkpoint file to purge since just creating a new TailPage from a HeadPage does
+                    // not trigger a checkpoint creation in itself
+                    TailPage tailPage = new TailPage(this.headPage);
+                    tailPage.purge();
+                } else {
+                    // beheading includes checkpoint+fsync if required
+                    TailPage tailPage = this.headPage.behead();
+
+                    this.tailPages.add(tailPage);
+                    if (! tailPage.isFullyRead()) {
```

Comment:
A page could be not fully acked but be fully read, where the read elements are awaiting to be acked.